### PR TITLE
[compiler-v2] Documentation for compiler beta

### PR DESCRIPTION
**Pull request by wrwg**
### Description

- Adds a new page `./move/tools/compiler_v2.mdx`
- Extends the existing function page in the Move book by documentation about receiver style calls

### Checklist

[None of the below commands work, various tools seem to be missing, spellcheck is unknown to pnpm]

- Do all Lints pass?
  - [] Have you ran `pnpm spellcheck`?
  - [] Have you ran `pnpm fmt`?
  - [] Have you ran `pnpm lint`?

### CODE REVIEW

**Review by gregnazario — APPROVED** [review]
Content lgtm!

File: apps/nextra/pages/en/build/move/compiler_v2.mdx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+title: "Compiler v2 (beta)"
+---
+
+# Move Compiler v2 (beta)
+
+Move compiler v2 is a new tool which translates Move source code into Move bytecode. It unifies the architectures of the Move Compiler and the Move Prover, enabling faster innovation in the Move language. It also offers new for defining code optimizations which can be leveraged to generate more gas efficient code for Move programs.
+
+## Beta Testing
+
+Compiler v2 (version `2.0-unstable`) is currently in beta testing and the developer community is encouraged to try it. However, the code produced by this version of the compiler should **NOT** be deployed on `mainnet`. Using it for unit tests or for a local nodes, `devnet`, and `testnet` is fine. This page will be updated once this changes.
```
Comment (rahxephon89) [inline apps/nextra/pages/en/build/move/compiler_v2.mdx:11]: Do we need a disclaimer for beta testing?

File: apps/nextra/pages/en/build/move.mdx [hunk 2]
```diff
@@ -23,3 +23,5 @@ The following external resources exist to further your education:
 - [IMCODING Move Tutorials](https://www.imcoding.online/tutorials?tag=Aptos)
 - [Pontem Move Playground](https://playground.pontem.network/)
 - [Collection of nestable Move resources](https://github.com/taoheorg/taohe)
+
+A new Move compiler and language version is currently in early beta testing. If you are interested to play with it, check [this page](./move/tools/compiler_v2.mdx).
```
Comment (rahxephon89) [inline apps/nextra/pages/en/build/move.mdx:27]: Is there a sub directory `tools` for `compiler_v2.mdx`? Maybe it is a github page issue but I cannot see it from this page.

File: apps/nextra/pages/en/build/move/book/functions.mdx [hunk 3]
```diff
@@ -24,24 +24,25 @@ Module functions, by default, can only be called within the same module. These i
...
-  }
+}
 
-  module other {
+module other {
     fun calls_m_foo(): u64 {
-      0x42::m::foo() // ERROR!
-//    ^^^^^^^^^^^^ 'foo' is internal to '0x42::m'
+        0x42::m::foo() // ERROR!
+        //    ^^^^^^^^^^^^ 'foo' is internal to '0x42::m'
```
Comment (vineethk) [inline apps/nextra/pages/en/build/move/book/functions.mdx:36]: The `^^^^` seems to be misaligned with this change. There are several instances of these in the file, better to grep and fix all of them.

File: apps/nextra/pages/en/build/move/book/functions.mdx [hunk 4]
```diff
@@ -609,3 +628,36 @@ There are plans to loosen some of these restrictions in the future, but for now,
...
+    //...
+
+    fun example() {
+        let s = S{};
+        s.foo(1);
+    }
+}
+```
+
+The call `s.foo(1)` is syntactic sugar for `f(&s, 1)`. Notice that the compiler automatically inserts the reference operator. The 2nd, old notation is still available for `f`, so one can incrementally introduce the new call style without breaking existing code.
```
Comment (vineethk) [inline apps/nextra/pages/en/build/move/book/functions.mdx:653]:
```suggestion
The call `s.foo(1)` is syntactic sugar for `foo(&s, 1)`. Notice that the compiler automatically inserts the reference operator. The 2nd, old notation is still available for `foo`, so one can incrementally introduce the new call style without breaking existing code.
```

File: apps/nextra/pages/en/build/move/book/functions.mdx [hunk 5]
```diff
@@ -609,3 +628,36 @@ There are plans to loosen some of these restrictions in the future, but for now,
...
+
+The type of the `self` argument can be a struct or an immutable or mutable reference to a struct. The struct must be declared in the same module as the function.
+
+Notice that you do not need to `use` the modules which introduce receiver functions. The compiler will find those functions automatically based on the argument type of `s` in a call like `s.foo(1)`. This, in combination with the automatic insertion of reference operators, can make code using this syntax significantly more concise.
+
+
```
Comment (vineethk) [inline apps/nextra/pages/en/build/move/book/functions.mdx:659]: Seems like a bunch of empty newlines that we would ideally get rid of.

File: apps/nextra/pages/en/build/move/compiler_v2.mdx [hunk 6]
```diff
@@ -0,0 +1,39 @@
+---
+title: "Compiler v2 (beta)"
+---
+
+# Move Compiler v2 (beta)
+
+Move compiler v2 is a new tool which translates Move source code into Move bytecode. It unifies the architectures of the Move Compiler and the Move Prover, enabling faster innovation in the Move language. It also offers new for defining code optimizations which can be leveraged to generate more gas efficient code for Move programs.
```
Comment (vineethk) [inline apps/nextra/pages/en/build/move/compiler_v2.mdx:7]: Could we use either "Move compiler" or "Move Compiler" consistently?

Comment (vineethk) [inline apps/nextra/pages/en/build/move/compiler_v2.mdx:7]: "It also offers new for defining" ....: There is a missing word or typo here.

**Review by vineethk — COMMENTED** [review]
Looking good so far.

File: apps/nextra/pages/en/build/move/compiler_v2.mdx (hunk 1 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move/compiler_v2.mdx:11]: I think what we have above is enough

File: apps/nextra/pages/en/build/move.mdx (hunk 2 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move.mdx:27]: Yes there is, at the given location. This links typically not work in PRs.

File: apps/nextra/pages/en/build/move/book/functions.mdx (hunk 3 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move/book/functions.mdx:36]: I think I fixed all instances. @gregnazario could that have come from some autoformatting you applied to the other example?

File: apps/nextra/pages/en/build/move/book/functions.mdx (hunk 4 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move/book/functions.mdx:653]: Done

File: apps/nextra/pages/en/build/move/book/functions.mdx (hunk 5 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move/book/functions.mdx:659]: Done

File: apps/nextra/pages/en/build/move/book/functions.mdx (hunk 3 above)
Comment (gregnazario) [reply apps/nextra/pages/en/build/move/book/functions.mdx:36]: I did them all manually, for some reason the prettier formatter destroys MDX files (we probably have to play with the settings)

File: apps/nextra/pages/en/build/move/compiler_v2.mdx (hunk 6 above)
Comment (wrwg) [reply apps/nextra/pages/en/build/move/compiler_v2.mdx:7]: Done

Comment (wrwg) [reply apps/nextra/pages/en/build/move/compiler_v2.mdx:7]: Done

**Review by hariria — APPROVED** [review]

**Review by vineethk — APPROVED** [review]

**Pull request merged**